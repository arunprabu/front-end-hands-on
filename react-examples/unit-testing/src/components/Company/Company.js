import React, { useState } from 'react'

const Company = (props) => {

  const [favCompany, setFavCompany] = useState('');

  const handleChange = (event) => {
    setFavCompany(event.target.value);
  }

  return (
    <div>
      <h2>Welcome to My Company's Official Webpage</h2>

      <input placeholder='Company Name' />

      <div data-testid="companyNameEl">
        {props.companyName}
      </div>

      <hr />

      <input
        data-testid="favCompanyNameInput"
        value={favCompany}
        onChange={handleChange} />
    </div>
  )
}

export default Company