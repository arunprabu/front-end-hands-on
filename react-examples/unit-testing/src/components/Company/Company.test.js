import { render, screen, cleanup, fireEvent } from '@testing-library/react';
import React from 'react'; // needed
import ReactDOM from 'react-dom'; // needed -- will use it to render the comp
import renderer from 'react-test-renderer';

import Company from "./Company";

afterEach(cleanup);

// test suite
describe('Company Component', () => {
  // test case #1
  it('renders the comp properly', () => {
    const divEl = document.createElement('div'); 
    ReactDOM.render(<Company />, divEl);
  });

  // test case #2 
  it('should have official company page as text', () => {
    render(<Company />);
    const officialTxtEl =screen.getByText('Welcome to My Company\'s Official Webpage');
    expect(officialTxtEl).toBeInTheDocument();
  });

  // test case #3 
  it('should have an input with Company Name as placeholder', ()=>{
    render(<Company />);
    const placeHolderText = screen.queryByPlaceholderText('Company Name');
    // if the placeholder is not matching, this will fail
    expect(placeHolderText).toBeTruthy();
  });

  // test case #4
  it('should display props correctly', () => {
    // passing prop to company compo
    render(<Company companyName='Twitter Inc.' />); 
    // locating an element with data-testid attribute with value companyNameEl
    const companyNameEl =  screen.getByTestId('companyNameEl'); 
    // checking if props received correctly
    expect(companyNameEl).toHaveTextContent('Twitter Inc.');
  });

  // let's check if the input element works well with onChange event
  it('should update on onChange event in company input', () => {
    // taking up the comp testing
    render(<Company />);
    // finding the element with test id companyNameInput
    const favCompanyNameInput = screen.getByTestId('favCompanyNameInput');
    // now checking if onChange is working or not
    // mock fire change event with the value
    fireEvent.change(favCompanyNameInput, {target: {value: 'test123'}});
    // finally checking if the input element is showing the right update
    expect(favCompanyNameInput.value).toBe('test123');
  });

  // snapshot testing
  it('should have right company comp snapshot', () => {
    const tree = renderer.create(<Company companyName='Twitter Inc.'/>).toJSON();
    console.log(tree);
    expect(tree).toMatchSnapshot();
  })
})