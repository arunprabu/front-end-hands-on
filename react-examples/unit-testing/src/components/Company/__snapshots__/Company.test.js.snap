// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Company Component should have right company comp snapshot 1`] = `
<div>
  <h2>
    Welcome to My Company's Official Webpage
  </h2>
  <input
    placeholder="Company Name"
  />
  <div
    data-testid="companyNameEl"
  >
    Twitter Inc.
  </div>
  <hr />
  <input
    data-testid="favCompanyNameInput"
    onChange={[Function]}
    value=""
  />
</div>
`;
